File: Backend/src/main/java/app/models/Product.java
package app.models;

import app.models.Project.Project;
import app.models.Project.ProjectType;

import javax.persistence.*;
import java.time.LocalDateTime;

@Entity
@Table(name = Product.TABLE_NAME)
public class Product implements Identifiable<Integer>
{

    public static final String TABLE_NAME = "product";

    @Id()
    @Column(name = "product_key", nullable = false, unique = true, updatable = false)
    @GeneratedValue(strategy = GenerationType.AUTO)
    private Integer id;

    @Column(name = "product_name", nullable = false)
    private String name;

    @Column(name = "description", nullable = true)
    private String description;

    @Column(name = "price", nullable = false, columnDefinition = "DECIMAL(10,2)")
    private double price;


    @Column(name = "product_stripe_id", nullable = true)
    private String stripe_id;

    @OneToOne(cascade = CascadeType.DETACH)
    @JoinColumn(name = "project_key", columnDefinition = "int", nullable = true,updatable = false,insertable = false)
    private Project project;

    @OneToOne(cascade = CascadeType.DETACH)
    @JoinColumn(name = "product_type_key", columnDefinition = "int", nullable = true,updatable = false,insertable = false)
    private ProjectType type;

    @Column(name = "created_at", nullable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_at", nullable = true)
    private LocalDateTime updatedAt;

    @Column(name = "product_id_ext", nullable = true)
    private Integer id_ext;

    @Column(name = "is_active", nullable = true)
    private boolean active;

    @Override
    public Integer getId() {
        return null;
    }

    @Override
    public void setId(Integer id) {

    }
}

File: frontend/src/Models/Product.js
import moment from "moment/moment";
import ProductType from "@/Models/ProductType";
import Project from "@/Models/Project";

export default class Product{
    id;
    name;
    description;
    price;
    project;
    type;
    createdAt;
    updatedAt;
    id_ext;
    active;
    constructor(id, name, description, price, project, type, createdAt, updatedAt, id_ext, active){
        this.id = id;
        this.name = name;
        this.description = description;
        this.price = price;
        this.project = project;
        this.type = type;
        this.createdAt = createdAt;
        this.updatedAt = updatedAt;
        this.id_ext = id_ext;
        this.active = active;
    }
    static copyConstructor(product){
        if(product == null) return null;
        product = Object.assign(new Product(),product);
        product.createdAt = moment(product.createdAt);
        product.updatedAt = moment(product.updatedAt);
        product.type = ProductType.copyConstructor(product.type);
        product.project = Project.copyConstructor(product.project);
        return product;
    }
}